File: src/main/java/org/codeberg/zenxarch/zombies/entity/effect/SwapPositionZombieEffect.java
package org.codeberg.zenxarch.zombies.entity.effect;

import com.mojang.serialization.MapCodec;
import net.minecraft.entity.LivingEntity;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.world.TeleportTarget;
import org.codeberg.zenxarch.zombies.entity.ExtendedZombieEntity;
import org.jetbrains.annotations.Nullable;

public record SwapPositionZombieEffect() implements ZombieEffect {
  private static final SwapPositionZombieEffect INSTANCE = new SwapPositionZombieEffect();
  public static final MapCodec<SwapPositionZombieEffect> CODEC = MapCodec.unit(INSTANCE);

  private static TeleportTarget getTeleportTarget(ServerWorld world, LivingEntity living) {
    return new TeleportTarget(
        world,
        living.getPos(),
        living.getVelocity(),
        living.getYaw(),
        living.getPitch(),
        TeleportTarget.NO_OP);
  }

  @Override
  public void run(
      ServerWorld world, ExtendedZombieEntity zombie, @Nullable LivingEntity adversery) {
    if (adversery == null) return;

    var toAdversary = getTeleportTarget(world, adversery);
    var toZombie = getTeleportTarget(world, zombie);

    zombie.teleportTo(toAdversary);
    adversery.teleportTo(toZombie);
  }

  @Override
  public MapCodec<SwapPositionZombieEffect> getCodec() {
    return CODEC;
  }

  public static SwapPositionZombieEffect create() {
    return INSTANCE;
  }
}

File: src/main/java/org/codeberg/zenxarch/zombies/entity/effect/SwapPositionZombieEffect.kt
package org.codeberg.zenxarch.zombies.entity.effect

import net.minecraft.entity.LivingEntity
import net.minecraft.server.world.ServerWorld
import net.minecraft.world.TeleportTarget
import org.codeberg.zenxarch.zombies.entity.ExtendedZombieEntity

class SwapPositionZombieEffect : ZombieEffect {
    override fun run(
        world: ServerWorld, zombie: ExtendedZombieEntity, adversery: LivingEntity?
    ) {
        if (adversery == null) return

        val toAdversary = getTeleportTarget(world, adversery)
        val toZombie = getTeleportTarget(world, zombie)

        zombie.teleportTo(toAdversary)
        adversery.teleportTo(toZombie)
    }

    companion object {
        private val INSTANCE = SwapPositionZombieEffect()

        private fun getTeleportTarget(world: ServerWorld, living: LivingEntity): TeleportTarget {
            return TeleportTarget(
                world,
                living.pos,
                living.velocity,
                living.yaw,
                living.pitch,
                TeleportTarget.NO_OP
            )
        }

        @JvmStatic
        fun create(): SwapPositionZombieEffect {
            return INSTANCE
        }
    }
}
